R/fragments-calculate2.R

```
@@ -0,0 +1,279 @@
+#' 
+#' @param sequence Character. A peptide sequence that may have modifications or not 
+#' 
+#' @param fixed_modifications Named numeric. Specifies which fixed modifications are used
+#' 
+#' @param variable_modifications Named numeric. Specifies which variable modifications are used
+#' 
+#' @param max_mods Numeric. Indicates how many modifications can be applied at once.
+#' 
+#' @return list with all possible combinations of modifications
+#' 
+#' @noRd
+#' 
+#' @examples
+#' .modificationPositions("ARGHKA", variable_modifications = c(A = 4, K = 5, S = 8), max_mods = 3)
+#' 
+.modificationPositions <- function(sequence,
+                                   variable_modifications = NULL,
+                                   max_mods = Inf) {
+    
+    sequence_split <- strsplit(sequence, "")[[1]]
+    
+    if (length(variable_modifications)) {
+        modifiable_positions_var <-
+            which(sequence_split %in% names(variable_modifications))
+        l <- length(modifiable_positions_var)
+    } else {l <- NULL}
+    
+    ## take the maximum amount of modifications possible 
+    max_mods <- min(max_mods, l)
+    
+    if (max_mods > 0 & length(l)) {
+        mod_combinations <- lapply(0:max_mods, function(n) {
+            if (n == 0) {
+                list(integer(0))
+            } else if (l == 1) {
+                list(modifiable_positions_var)
+            } else {
+                combn(modifiable_positions_var, n, simplify = FALSE)
+            }
+        })
+        
+        mod_combinations <- do.call(c, mod_combinations)
+        lapply(mod_combinations, function(comb) {
+            mods <- rep(0, nchar(sequence))
+            mods[comb] <- unname(variable_modifications[sequence_split[comb]])
+            names(mods) <- sequence_split
+            mods
+        })
+    } else {
+        mods <- rep(0, nchar(sequence))
+        names(mods) <- sequence_split
+        list(mods)
+    }
```

This is the same return value for `max_mod < 1` and `!length(variable_modifications)`. I would put it at the beginning of the function to fail early and keep the function simple.